### PR TITLE
patch: add support for changing directory

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Example: Homebrew/homebrew-core#51129 could be rewritten as

```rb
patch do
  url "https://github.com/boostorg/boost_install/compare/52ab9149544bae82e54f600303f5d6d1dda9a4f5...a1b5a477470ff9dc2e00f30be4ec4285583b33b6.patch?full_index=1"
  sha256 "fb168dd2ddfa20983b565ead86d4355c6d6e3e49bce9c2c6ab7f6e9cd9350bd4"
  directory "tools/boost_install"
end
```

### CODE REVIEW

**Comment by bayandin** [comment]
Btw, you can apply a patch for a resource by adding the patch block to resource block, for example https://github.com/Homebrew/homebrew-core/blob/eb246b43ff3fec92fbdeaa615db5f988cf87cf50/Formula/azure-cli.rb#L486-L495

If this is possible to do it looks more preferable for me since it keeps related things together (especially useful if you have a lot of resources)

**Comment by bayandin** [comment]
Ah, I misread the example, my comment above completely unrelated for this case, sorry for the noise 🤷‍♂

**Comment by Bo98** [comment]
Yeah, the Boost example is perhaps a tad confusing - which is probably a good reason to incorporate proper patch subdirectory support!

File: Library/Homebrew/patch.rb [hunk 1]
```diff
@@ -149,6 +149,8 @@ def apply
 
         patch_files << children.first.basename
       end
+      dir = base_dir
+      dir /= resource.directory unless resource.directory.nil?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/patch.rb]:
```suggestion
      dir /= resource.directory if resource.directory.present?
```

**Comment by Bo98** [comment]
Done.

**Comment by MikeMcQuaid** [comment]
@Bo98 GitHub Actions being weird; can you force push this? Ta!

**Comment by MikeMcQuaid** [comment]
Nice work again @Bo98!

**Pull request merged**